### PR TITLE
Make `mark` consistent with `skip` and `IOBuffer`

### PR DESCRIPTION
**Pull request by nhz2**

### CODE REVIEW

File: ext/TestExt.jl [hunk 1]
```diff
@@ -88,11 +88,15 @@ function TranscodingStreams.test_chunked_read(Encoder, Decoder)
         ok = true
         for chunk in chunks
             stream = TranscodingStream(Decoder(), buffer, stop_on_end=true)
-            ok &= hash(read(stream)) == hash(chunk)
+            ok &= read(stream) == chunk
             ok &= eof(stream)
             ok &= isreadable(stream)
             close(stream)
         end
+        # read without stop_on_end should read the full data.
```
Comment (nhz2) [inline ext/TestExt.jl:96]: This test isn't related to this PR but was something I noticed wasn't covered by current tests.

File: src/stream.jl [hunk 2]
```diff
@@ -229,50 +229,26 @@ end
...
     checkmode(stream)
-    return ismarked(stream.buffer1)
+    isopen(stream) && ismarked(stream.buffer1)
 end
 
-function Base.mark(stream::TranscodingStream)
-    checkmode(stream)
-    return mark!(stream.buffer1)
+function Base.mark(stream::TranscodingStream)::Int64
+    ready_to_read!(stream)
```
Comment (nhz2) [inline src/stream.jl:238]: This throws an error if `mark` is used on a stream in write, close, or panic modes.

File: src/stream.jl [hunk 3]
```diff
@@ -229,50 +229,26 @@ end
...
-    if offset < 0
-        throw(ArgumentError("negative offset"))
-    end
-    mode = stream.state.mode
-    buffer1 = stream.buffer1
-    skipped = 0
-    if mode === :read || mode === :stop
-        while !eof(stream) && buffersize(buffer1) < offset - skipped
-            n = buffersize(buffer1)
-            emptybuffer!(buffer1)
```
Comment (nhz2) [inline src/stream.jl:263]: The use of `emptybuffer!` here was discarding potentially marked data.

File: src/stream.jl [hunk 4]
```diff
@@ -229,50 +229,26 @@ end
...
-            skipped += n
-        end
-        if eof(stream)
-            emptybuffer!(buffer1)
-        else
-            skipbuffer!(buffer1, offset - skipped)
-        end
-    else
-        # TODO: support skip in write mode
```
Comment (nhz2) [inline src/stream.jl:272]: I'm not sure what this TODO means, but I add a docstring to clarify that `skip` behaves like a read that ignores `eof` and discards the read bytes.

Comment (nhz2) [reply src/stream.jl:272]: I also moved the `skip` to be near the other read functions

**Review by mkitti — APPROVED** [review]
Looks good to me.

**Pull request merged**